### ISSUE TITLE
Use curses everywhere for a nice text user interface

### ISSUE THREAD

**Issue by sophron** [wifiphisher/wifiphisher#611]
We are currently using:

1) `curses` for targeting an Access Point, 
2) `blessings` (`curses` wrapper) for our main view.
3)  custom terminal interface for selecting a template

`curses` is the standard library for TUI and we should use it in all three above cases. 

@blackHatMonkey: You did a great job for targeting an AP. Feel free to assign yourself if you are planning to work on this. :)

**Comment by blackHatMonkey** [comment]
If we want to  move all the code to `curses` we first need to come up with some generic methods for printing and keep track of all the stuff. This is crucial since the amount code just for the access point selection is huge and adding more sections to it will just not work.

**Comment by anakin1028** [comment]
@blackHatMonkey  can I sidetrack ask when we use `curses` how do you use `pdb` to debug? Since they seem to interfere each other right?

**Comment by blackHatMonkey** [comment]
One of the very first bugs that you caught was in the curses section and I was not able to use `pdb` on it. What I did was just test the logic of it rather than the whole code.

**Comment by anakin1028** [comment]
Yes at that time I can just write test code to see if my solution is work.

**Comment by sophron** [comment]
@blackHatMonkey Yes, we should have our own separate module `tui.py` for anything regarding `curses`.

**Comment by blackHatMonkey** [comment]
@sophron @anakin1028  After thinking it this for some time I don't this is a good idea. Some of the reasons are:
- it is impossible to debug the code inside the `curses` wrapper
- every time consuming
- very ugly as you need functions with huge amount code in there
- very hard if not impossible to test (writing test cases)

What I purpose is making everything using the built in print and essentially dumb it down to the bare minimums. After that we add the GUI from the browser which would be much easier to design and handle.

**Comment by sophron** [comment]
>  it is impossible to debug the code inside the `curses` wrapper

Indeed, there are many C-like calls that make a mess.

>  every time consuming

I'm not sure I understand this argument.

>  very ugly as you need functions with huge amount code in there

Yes, it's not the most Pythonic library.

> very hard if not impossible to test (writing test cases)

I believe it is hard to write test cases for any TUI library.

> What I purpose is making everything using the built in print and essentially dumb it down to the bare minimums. After that we add the GUI from the browser which would be much easier to design and handle.

I am fine with `print`. However, when we want to display states (e.g. during AP selection or in our main view), `print` by itself won't help us. What we are doing now is a hacking way of mixing `print` with `clear` in a `while` loop. I find this ugly and `curses` is the only way to display this information nicely from within the terminal. 

I agree that `curses` is not Pythonic but it's also our only choice for a good terminal experience. And a good terminal experience is a priority (people may want to run the tool in embedded devices with no installed GUI).

**Comment by anakin1028** [comment]
Do we stick to use `curses`? From my points `curses` makes debug very difficult like the issues we have before `rssi`, `malformed beacon` are all in the `curses` section. 

However, I agree with @sophron's point we need to consider the `embedded devices' case`.
Anyway, I'm doing some study about curses now, and I can change the main terminal from `blessed` to `curses`.  Like the below figure:

![curses](https://user-images.githubusercontent.com/6182530/27190467-206ede76-5227-11e7-8605-85ad19d5e15b.png)

The refactor process may be done in the next week I think if we decide to stick to `curses`.

**Comment by sophron** [comment]
Yes, I believe we should stick to `curses` for the reasons I mentioned.

Grood job, that looks good. We also need to move the "template selection" to curses too.

**Comment by blackHatMonkey** [comment]
One suggestion I would make if you are going to implement it there would be to move the info about the `ESSID` and others to the top of the page rather than top right corner. So the result would be something similar

```
ESSID: whatever
Channel: 1
...


Deauthenticating clients:
...

DHCP Leases:
....

HTTP requests:
....


```

**Comment by anakin1028** [comment]
@blackHatMonkey Ok I'll adjust this : )

**Comment by sophron** [comment]
@blackHatMonkey Why do you want it moved to the top?

The most important thing here is to try and fit everything into 80x24 characters to support VT100-like terminals. Curses allows us to be flexible and save some "unused" space with printing information regarding the version and the running AP.

If we are looking for improvements, there are many things that we can do to make it look better. There's no need to <a href="http://www.tldp.org/LDP/lpg/node104.html">draw lines</a> by concatenating dashes. For more ideas, have a look at Kismet which its default UI uses ncurses.
